nucypher/blockchain/economics.py

```diff
@@ -110,88 +244,95 @@ def __init__(self,
-            self.locked_periods_coefficient,  # Locked periods coefficient (k1)
-            self.maximum_locked_periods,      # Max periods that will be additionally rewarded (awarded_periods)
-
-            # Constraints
-            self.minimum_locked_periods,      # Min amount of periods during which tokens can be locked
-            self.minimum_allowed_locked,      # Min amount of tokens that can be locked
-            self.maximum_allowed_locked,      # Max amount of tokens that can be locked
-            self.minimum_worker_periods       # Min amount of periods while a worker can't be changed
+        super(StandardTokenEconomics, self).__init__(
+            initial_supply,
+            total_supply,
+            staking_coefficient,
+            locked_periods_coefficient,
+            maximum_rewarded_periods,
+            **kwargs
         )
-        return tuple(map(int, deploy_parameters))
-
-
-class SlashingEconomics:
 
-    HASH_ALGORITHM_KECCAK256 = 0
-    HASH_ALGORITHM_SHA256 = 1
-    HASH_ALGORITHM_RIPEMD160 = 2
+    def token_supply_at_period(self, period: int) -> int:
+        if period < 0:
+            raise ValueError("Period must be a positive integer")
 
-    hash_algorithm = HASH_ALGORITHM_SHA256
-    base_penalty = 100
-    penalty_history_coefficient = 10
-    percentage_penalty_coefficient = 8
-    reward_coefficient = 2
-
-    @property
-    def deployment_parameters(self) -> Tuple[int, ...]:
-        """Cast coefficient attributes to uint256 compatible type for solidity+EVM"""
-
-        deployment_parameters = [
-            self.hash_algorithm,
-            self.base_penalty,
-            self.penalty_history_coefficient,
-            self.percentage_penalty_coefficient,
-            self.reward_coefficient
-        ]
+        with localcontext() as ctx:
+            ctx.prec = self._precision
 
-        return tuple(map(int, deployment_parameters))
+            # Eq. 3 of the mining paper
+            # https://github.com/nucypher/mining-paper/blob/master/mining-paper.pdf
+            t = Decimal(period)
+            S_0 = self.erc20_initial_supply
+            i_0 = 1
+            I_0 = i_0 * S_0  # in 1/years
+            T_half = self.token_halving  # in years
+            T_half_in_days = T_half * 365
+
+            S_t = S_0 + I_0 * T_half * (1 - 2**(-t / T_half_in_days)) / LOG2
+            return int(S_t)
+
+    def cumulative_rewards_at_period(self, period: int) -> int:
+        return self.token_supply_at_period(period) - self.erc20_initial_supply
+
+    def rewards_during_period(self, period: int) -> int:
+        return self.cumulative_rewards_at_period(period) - self.cumulative_rewards_at_period(period-1)
+
+
+class TokenEconomicsFactory:
+    # TODO: Enforce singleton
+
+    __economics = dict()
+
+    @classmethod
+    def get_economics(cls, registry: BaseContractRegistry) -> TokenEconomics:
+        registry_id = registry.id
```

Comment:
Yeah, interesting.  This means only one token economics class per deployment set, per runtime, which more tightly couples the idea of  ` deployment_set <-> registry <-> agent`.  My first thought was that this will prevent holding two agencies at the same time with different economics, but then it occurred to me that *different economics == different registry*.